client.man
.\"$Id$
.TH cfclient "November 29, 1998"
.SH NAME
cfclient - X11 client program for crossfire servers.
.br
gcfclient - gtk client program for crossfire servers.
.SH SYNOPSIS
.B client [-server server] [-port number] [-display display] [-split|-nosplit]
[-echo] [-pix|-xpm] [-showicon] [-scrollines number] [-sync] [-help]
[-cache|-nocache]
.SH DESCRIPTION
.PP
.I Cfclient/gcflient
is an X11 program for connecting to crossfire servers.  Crossfire is a
multiplayer graphical adventure game that mixes aspects of Gauntlet (TM),
nethack, moria, and Ultima (TM).

This man page describes options for the X11 client of crossfire
(cfclient).  Many of these options should also be valid for the gtk
version (gcfclient).  The configure/make process will build the
gtk version if it finds the necessary libraries, otherwise it will
build the X11 version.

.SH OPTIONS
.TP
.B -server <server>
Sets the server to play on.  If no option is given, it will default to
localhost.  
.TP
.B -port <number>
The port number of the server to connect to.  By default, it will use
port 13327 (standard port).  However, certain servers may use other
ports.
.TP
.B -display <display>
The display to open the windows on.  If this option is not given, it
will use the DISPLAY environmental variable to determine the display.
.TP
.B -split|-nosplit
-split starts the windows up in split windows mode.  In split windows
mode, 6 seperate windows are created - these individual windows can then
be moved around and resized as desired.  -nosplit starts the game up with
a single window - this is the default.  The option is useful if your
saved defaults are set for -split (see Saved Defaults further down).

.TP
.B -echo
Echo commands as they are entered.  Normally, commands bound to keys
are just sent to the server without any echoing on the client of what
that command actually was.  This option causes the commands to also be
printed in the information window as they are sent to the server.

.TP
.B -pix|-xpm
This determines the graphic image types to use.  -pix uses pixmap images -
these are two color images with no masking (one image per space.).  -xpm
uses XPM format images - these are multi colored images with masking,
so multiple images per space can be displayed.  XPM images require
more local processing power (unless you are on a very old machine, this
should not be an issue), the installation of the XPM library, and also
require more bandwidth to download (however, see the -cache option
further down.)

.TP
.B -showicon
This shows a little icon next to items in your inventory that contains
a brief description of some of the item properties (magic, cursed,
equipped, etc.)  This can make spotting some items easier, but some
players may not like the extra space these icons take up or the

.TP
.B -scollines <lines>
This is the number of lines to use in the information window.  By default,
the size of the window is used (no scrollback ability).  With this
option, a scrollbar will be created and it is then possible to scroll back
through old messages.  It is strongly recommended you set this to some
value, since some areas output more data than will fit in the output
window at one time.

.TP
.B -sync
Runs the server in synchronous display mode.  This option tends only to
be useful in debugging purposes - using this will slow down the display
and not gain anything for the typical player.

.TP
.B -help
Prints out a brief description of the options to standard output.

.TP
.B -cache|-nocache
Determines if the client will cache images for future runs.  With -nocache,
when the server is about to referance an image that the client has
yet to receive, it sends that image along to the client.  These downloaded
images are discarded once the client exists.  With -cache, the server
sends information about the image it is about to referance.  The client
can then check its local cache (stored in ~/.crossfire/images) - if it
has the image, it creates it off the disk.  If it does not have it,
it requests it from the server and then stores it in its cache.  In the
brief time between the request of the image and it actually receives it,
that image will be represented by a ? in the map.

-cache option is strongly recommended anytime when your link to the server
is slower than home directory access - this is likely to be the case
except in case of nfs mounted home directories on the server on the local
lan.
.TP
.B -updatekeycodes
The standard behaviour when a player uses the bind command to bind
new actions is that they keycode is saved with that binding.  Keycodes
are specific to keyboards - a sun keyboard will generate a different
keycode compared to a PC style keyboard.  In most cases, it is not
always desirable to have the keycodes get updated, as this may make
some bindings unavailable.  Using this option will force custom keybindings
to get updated for the current keyboard when the client is run.

.SH SAVED DEFAULTS

 Once you have logged into the server, you can enter extended commands
by typing ' (apostrophe) followed by the extended commands.  One of these
extended commands is savedefaults.  This will save the server, port,
image type, split windows mode, scrolllines, showicon, and cache options into
~/.crossfire/defaults.

If you are using split windows mode, you can enter savewinpos
as an extended command.  This will save the current window positions
and sizes into ~/.crossfire/winpos.  Next time you run the server
in split mode, it will load these values and resize and move the windows
as specified.

.SH PLAYING WITH THE CLIENT

This is by far a complete tutorial of gameplay.  This section is to
provide a basic introduction to logging in and doing a few very
basic actions.

The first thing that will happen after the client successfully connects
to the server is you will be prompted for a name.  There are some
restrictions to the name you can choose, but alphanumeric characters
are all safe.

After entering the name, you will then be prompted for a password.  When
creating a new character, enter whatever you want your password to be.
If you get a 'login incorrect' someone else already has that character
name with a different password - try another name.  If no one has used
that name, you will be prompted for the password again to confirm it.

You now start the character creation process.  It is beyond the scope
of the document to fully describe the pros and cons of different
stats and classes.  The creation process has different options for
swapping stats and choosing a class.

Once you have chosen your class, you will be in the middle of a town.
The arrow keys will move you in the various directions.  There will be
a red building almost directly above you - this is the inn where you
save your character.  There are some signs to your right.  To read
them, move on top of them and press the 'a' key.  This applies the
object below you - in the case of signs, it reads it.  In the case of
buildings, it will cause you to enter the building.

To the left of the signs is a small building - this is the beginner
dungeon.  Enter it, and follow the directions on the signs.  It provides
a pretty good tutorial on some of the basic actions and features of
the game.

To enter extended commands, type the ' (apostrophe) followed by the
command.  An example would be 'help (apostrophe followed by the
help command).  Enter command to execute the command.

.SH USEFUL EXTENDED COMMANDS

.TP
.B savewinpos savedefaults
These commands were described in the SAVED DEFAULTS options above.

.TP
.B scroll
This toggles whether or the information windows scrolls when it gets to 
the bottom of the window or wraps to the top.  Wrapping is slighly less
cpu intensive, but is generally harder to read.

.TP
.B bind unbind
bind is used to add new keybindings.  Do you want to be able to press
one key to cast a fireball?  This is what the bind command does.
'help bind gives much more detailed information.  Once a command
is bound, it is stored in ~/.crossfire/keys and will be used in future
plays.

.SH FILES
.TP
.B ~/.crossfire
This is a directory that is created that is used to store various
files.
.TP
.B ~/.crossfire/defaults
This contains default setttings.  This file can in theory be edited
by hand - just be careful to keep the same case and whitespace.
.TP
.B ~/.crossfire/images
This directory is used for the image cache if -cache is being used.
.TP
.B ~/.crossfire/keys
This contains keybindings you have edited.  This file can also be
theoretically edited - changing existing entries is much safer than
trying to figure out the syntax for new entries.
.TP
.B ~/.crossfire/winpos
Window positions and sizes to use in split window mode.  This file
can also be edited by hand.

.PP
Please let me know about any bugs you find in the client.
.SH AUTHOR
Copyright (C) 1994,1996,1998 Mark Wedel (mark@pyramid.com)
GTK port by David Sundqvist (azzie@netpolicy.com)

There are a great many other contributors to both the client and server
that are not mentioned here.

.ft R
.RE
.LP
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.

.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.

.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

.\" The author can be reached via e-mail to frankj@ifi.uio.no.


NEWS
This file is no longer maintained; information about relevant changes can be
found in the ChangeLog file.

This file is here largely because it is required by automake.


client.c
 /* Client interface main routine.
  * this file sets up a few global variables, connects to the server,
  * tells it what kind of pictures it wants, adds the client and enters
  * the main dispatch loop
  *
  * the main event loop (event_loop()) checks the tcp socket for input and
  * then polls for x events.  This should be fixed since you can just block
  * on both filedescriptors.
  *
  * The DoClient function recieves a message (an ArgList), unpacks it, and
  * in a slow for loop dispatches the command to the right function through
  * the commands table.   ArgLists are essentially like RPC things, only 
  * they don't require going through RPCgen, and it's easy to get variable
  * length lists.  They are just lists of longs, strings, characters, and 
  * byte arrays that can be converted to a machine independent format
 */


#include <client.h>

/* actually declare the globals */

char *server=SERVER,*client_libdir=NULL,*meta_server=META_SERVER;
char *image_file="";

int port_num=EPORT, meta_port=META_PORT, want_skill_exp=0, mapx=11, mapy=11,
    want_mapx=11, want_mapy=11;
FILE *fpin,*fpout;
int fdin, fdout, basenrofpixmaps, pending_images=0,maxfiledescriptor,
	pending_archs=0,maxfd,map1cmd=0,metaserver_on=METASERVER;
Client_Player cpl;
ClientSocket csocket;

char *resists_name[NUM_RESISTS] = {
"armor", "magic", "fire", "elec", 
"cold", "conf", "acid", "drain",
"ghit", "pois", "slow", "para",
"t undead", "fear", "depl","death", 
"hword", "blind"};

char *skill_names[MAX_SKILL] = {
"agility", "personality", "mental", "physique", "magic", "wisdom"
};

typedef void (*CmdProc)(unsigned char *, int len);

struct CmdMapping {
  char *cmdname;
  void (*cmdproc)(unsigned char *, int );
};


struct CmdMapping commands[] = {
    /* Order of this table doesn't make a difference.  I tried to sort
     * of cluster the related stuff together.
     */
    { "map", MapCmd },
    { "map1", Map1Cmd },
    { "map_scroll", (CmdProc)map_scrollCmd },
    { "magicmap", MagicMapCmd},

    { "item", ItemCmd },
    { "item1", Item1Cmd },
    { "upditem", UpdateItemCmd },
    { "delitem", DeleteItem },
    { "delinv",	DeleteInventory },

    { "drawinfo", (CmdProc)DrawInfoCmd },
    { "stats", StatsCmd },

    { "pixmap", PixMapCmd },
    { "bitmap", BitMapCmd },
    { "image", ImageCmd },
    { "face", FaceCmd},
    { "face1", Face1Cmd},


    { "sound", SoundCmd},
    { "anim", AnimCmd},

    { "player", PlayerCmd },
    { "comc", CompleteCmd},

    { "addme_failed", (CmdProc)AddMeFail },
    { "addme_success", (CmdProc)AddMeSuccess },
    { "version", (CmdProc)VersionCmd },
    { "goodbye", (CmdProc)GoodbyeCmd },
    { "setup", (CmdProc)SetupCmd},

    { "query", (CmdProc)handle_query},
};

#define NCOMMANDS (sizeof(commands)/sizeof(struct CmdMapping))

void DoClient(ClientSocket *csocket)
{
    int i,len;
    unsigned char *data;

    while (1) {
	i=SockList_ReadPacket(csocket->fd, &csocket->inbuf, MAXSOCKBUF-1);
	if (i==-1) {
	    /* Need to add some better logic here */
	    fprintf(stderr,"Got error on read (error %d)\n", errno);
	    csocket->fd=-1;
	    return;
	}
	if (i==0) return;   /* Don't have a full packet */
	csocket->inbuf.buf[csocket->inbuf.len]='\0';
        data = (unsigned char *)strchr((char*)csocket->inbuf.buf +2, ' ');
	if (data) {
	    *data='\0';
	    data++;
	}
        len = csocket->inbuf.len - (data - csocket->inbuf.buf);
	/* Terminate the buffer */
	LOG(0,"Command:%s (%d)\n",csocket->inbuf.buf+2, len);
	for(i=0;i < NCOMMANDS;i++) {
	    if (strcmp((char*)csocket->inbuf.buf+2,commands[i].cmdname)==0) {
		    commands[i].cmdproc(data,len);
		    break;
	    }
	}
	csocket->inbuf.len=0;
	if (i == NCOMMANDS) {
	    printf("Bad command from server (%s)\n",csocket->inbuf.buf+2);
	}
    }
}

#include <netdb.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <ctype.h>
#include <arpa/inet.h>

/* returns the fd of the connected socket, -1 on failure. */

int init_connection(char *host, int port)
{
    struct protoent *protox;
    int fd, oldbufsize, newbufsize=65535, buflen=sizeof(int);
    struct sockaddr_in insock;

    protox = getprotobyname("tcp");
    if (protox == (struct protoent  *) NULL)
    {
	fprintf(stderr, "Error getting prorobyname (tcp)\n");
	return -1;
    }
    fd = socket(PF_INET, SOCK_STREAM, protox->p_proto);
    if (fd==-1) {
	perror("init_connection:  Error on socket command.\n");
	return -1;
    }
    insock.sin_family = AF_INET;
    insock.sin_port = htons((unsigned short)port);
    if (isdigit(*host))
	insock.sin_addr.s_addr = inet_addr(host);
    else {
	struct hostent *hostbn = gethostbyname(host);
	if (hostbn == (struct hostent *) NULL)
	{
	    fprintf(stderr,"Unknown host: %s\n",host);
	    return -1;
	}
	memcpy(&insock.sin_addr, hostbn->h_addr, hostbn->h_length);
    }
    if (connect(fd,(struct sockaddr *)&insock,sizeof(insock)) == (-1))
    {
	perror("Can't connect to server");
	return -1;
    }
    if (fcntl(fd, F_SETFL, O_NDELAY)==-1) {
	fprintf(stderr,"InitConnection:  Error on fcntl.\n");
    }

    if (getsockopt(fd,SOL_SOCKET,SO_RCVBUF, (char*)&oldbufsize, &buflen)==-1)
        oldbufsize=0;

    if (oldbufsize<newbufsize) {
	if(setsockopt(fd,SOL_SOCKET,SO_RCVBUF, (char*)&newbufsize, sizeof(&newbufsize))) {
            LOG(1,"InitConnection: setsockopt unable to set output buf size to %d\n", newbufsize);
	    setsockopt(fd,SOL_SOCKET,SO_RCVBUF, (char*)&oldbufsize, sizeof(&oldbufsize));
	}
    }
    return fd;
}

/* This function negotiates/establishes the connection with the
 * server.
 */

void negotiate_connection(int sound)
{

    int cache;
    char buf[BIG_BUF];

    SendVersion(csocket);

    /* We need to get the version command fairly early on because
     * we need to know if the server will support a request to use
     * png images.  This isn't done the best, because if the server
     * never sends the version command, we can loop here forever.
     * However, if it doesn't send the version command, we have no idea
     * what we are dealing with.
     */
    while (csocket.cs_version==0) {
	DoClient(&csocket);
    }


    cache = display_willcache();
    if (cache) cache = CF_FACE_CACHE;

    if (display_usebitmaps()) 
	SendSetFaceMode(csocket,CF_FACE_BITMAP | cache); 
    else if (display_usexpm()) 
	SendSetFaceMode(csocket,CF_FACE_XPM | cache);
    else if (display_usepng()) {
	if (csocket.sc_version<1023) {
	    SendSetFaceMode(csocket,CF_FACE_XPM | cache);
	    draw_info("Server does not support PNG images.  Will use XPM instead", NDI_RED);
	}
	else SendSetFaceMode(csocket,CF_FACE_PNG | cache);
    }

# if 0
    if (sound<0)
	cs_write_string(csocket.fd,"setsound 0", 10);
    else
	cs_write_string(csocket.fd,"setsound 1", 10);
#else
    sprintf(buf,"setup sound %d sexp %d",
	    (sound<0?0:1), want_skill_exp);
    cs_write_string(csocket.fd, buf, strlen(buf));

#endif
    mapx=11;
    mapy=11;
    if (want_mapx!=11 || want_mapy!=11) {
	sprintf(buf,"setup mapsize %dx%d",want_mapx, want_mapy);
	cs_write_string(csocket.fd, buf, strlen(buf));
    }

    SendAddMe(csocket);
}


int main(int argc, char *argv[])
{
    int sound,got_one=0;

    /* This needs to be done first.  In addition to being quite quick,
     * it also sets up some paths (client_libdir) that are needed by
     * the other functions.
     */

    init_client_vars();
    
    /* Call this very early.  It should parse all command
     * line arguments and set the pertinent ones up in
     * globals.  Also call it early so that if it can't set up
     * the windowing system, we get an error before trying to
     * to connect to the server.  And command line options will
     * likely change on the server we connect to.
     */
    if (init_windows(argc, argv)) {	/* x11.c */
	fprintf(stderr,"Failure to init windows.\n");
	exit(1);
    }
    csocket.inbuf.buf=malloc(MAXSOCKBUF);

#ifdef HAVE_SYSCONF
    maxfd = sysconf(_SC_OPEN_MAX);
#else
    maxfd = getdtablesize();
#endif

    sound = init_sounds();

    /* Loop to connect to server/metaserver and play the game */
    while (1) {
	reset_client_vars();
	csocket.inbuf.len=0;
	csocket.cs_version=0;

	/* Perhaps not the best assumption, but we are taking it that
	 * if the player has not specified a server (ie, server
	 * matches compiled in default), we use the meta server.
	 * otherwise, use the server provided, bypassing metaserver.
	 * Also, if the player has already played on a server once (defined
	 * by got_one), go to the metaserver.  That gives them the oppurtunity
	 * to quit the client or select another server.  We should really add
	 * an entry for the last server there also.
	 */

	if (!strcmp(server, SERVER) || got_one) {
	    char *ms;
	    metaserver_get_info(meta_server, meta_port);
	    metaserver_show(TRUE);
	    do {
		ms=get_metaserver();
	    } while (metaserver_select(ms));
	    negotiate_connection(sound);
	} else {
	    csocket.fd=init_connection(server, port_num);
	    if (csocket.fd == -1) { /* specified server no longer valid */
		server = SERVER;
		continue;
	    }
	    negotiate_connection(sound);
	}

	got_one=1;
	event_loop();
	/* if event_loop has exited, we most of lost our connection, so we
	 * loop again to establish a new one.
	 */

	/* Need to reset the images so they match up properly and prevent
	 * memory leaks.
	 */
	reset_image_data();
	remove_item_inventory(cpl.ob);
	/* We know the following is the private map structure in
	 * item.c.  But we don't have direct access to it, so
	 * we still use locate.
	 */
	remove_item_inventory(locate_item(0));
    }
    exit(0);	/* never reached */
}
